File: src/main/java/com/github/grishberg/cad3d/keyboard/cfg/KeyboardConfig.java
package com.github.grishberg.cad3d.keyboard.cfg;


public class KeyboardConfig {

    private final int fn;
    private final double plateZOffset;
    private final double rowCurvature;
    private final double tentingAngle;
    private final double columnCurvature;
    private final double keyswitchHeight;
    private final double keyswitchWidth;
    private final double extraWidth;
    private final double extraHeight;
    private final double plateThickness;
    private final double saProfileKeyHeight;
    private final int centerCol;
    private final int rowsCount;
    private final int columnsCount;
    private final int centerRow;
    private final double keyPlaceHolderWidth;
    private final double keyPlaceHolderDepth;
    private final double keyPlaceHolderHeight;
    private final boolean isLowProfile;
    private final KeyZAngleProvider zAngleProvider;
    private final KeyOffsetProvider columnOffsetProvider;
    private final PowerSwitcherType powerSwitcherType;
    private final boolean hasHotswap;
    private final boolean magneticWristRestHolder;

    private final double bordersOffset;

    public KeyboardConfig(
        int fn,
        double plateZOffset,
        double columnCurvature,
        double tentingAngle,
        double rowCurvature,
        double keyswitchHeight,
        double keyswitchWidth,
        double extraWidth,
        double extraHeight,
        double plateThickness,
        double saProfileKeyHeight,
        int centerCol,
        int rowsCount,
        int columnsCount,
        int centerRow,
        double keyPlaceHolderWidth,
        double keyPlaceHolderDepth,
        double keyPlaceHolderHeight,
        boolean isLowProfile,
        KeyZAngleProvider zAngleProvider,
        KeyOffsetProvider columnOffsetProvider,
        PowerSwitcherType powerSwitcherType,
        boolean hasHotswap,
        boolean magneticWristRestHolder,
        double bordersOffset
    ) {
        this.fn = fn;
        this.plateZOffset = plateZOffset;
        this.rowCurvature = columnCurvature;
        this.tentingAngle = tentingAngle;
        this.columnCurvature = rowCurvature;
        this.keyswitchHeight = keyswitchHeight;
        this.keyswitchWidth = keyswitchWidth;
        this.extraWidth = extraWidth;
        this.extraHeight = extraHeight;
        this.plateThickness = plateThickness;
        this.saProfileKeyHeight = saProfileKeyHeight;
        this.centerCol = centerCol;
        this.rowsCount = rowsCount;
        this.columnsCount = columnsCount;
        this.centerRow = centerRow;
        this.keyPlaceHolderWidth = keyPlaceHolderWidth;
        this.keyPlaceHolderDepth = keyPlaceHolderDepth;
        this.keyPlaceHolderHeight = keyPlaceHolderHeight;
        this.isLowProfile = isLowProfile;
        this.zAngleProvider = zAngleProvider;
        this.columnOffsetProvider = columnOffsetProvider;
        this.powerSwitcherType = powerSwitcherType;
        this.hasHotswap = hasHotswap;
        this.magneticWristRestHolder = magneticWristRestHolder;
        this.bordersOffset = bordersOffset;
    }

    public int getFn() {
        return fn;
    }

    public double getPlateZOffset() {
        return plateZOffset;
    }

    public double getRowCurvature() {
        return rowCurvature;
    }

    public double getTentingAngle() {
        return tentingAngle;
    }

    public double getColumnCurvature() {
        return columnCurvature;
    }

    public double getKeyswitchHeight() {
        return keyswitchHeight;
    }

    public double getKeyswitchWidth() {
        return keyswitchWidth;
    }

    public double getExtraWidth() {
        return extraWidth;
    }

    public double getExtraHeight() {
        return extraHeight;
    }

    public double getPlateThickness() {
        return plateThickness;
    }

    public double getSaProfileKeyHeight() {
        return saProfileKeyHeight;
    }

    public int getCenterCol() {
        return centerCol;
    }

    public int getRowsCount() {
        return rowsCount;
    }

    public int getColumnsCount() {
        return columnsCount;
    }

    public int getCenterRow() {
        return centerRow;
    }

    public double getKeyPlaceHolderWidth() {
        return keyPlaceHolderWidth;
    }

    public double getKeyPlaceHolderDepth() {
        return keyPlaceHolderDepth;
    }

    public double getKeyPlaceHolderHeight() {
        return keyPlaceHolderHeight;
    }

    public boolean isLowProfile() {
        return isLowProfile;
    }

    public KeyZAngleProvider getZAngleProvider() {
        return zAngleProvider;
    }

    public KeyOffsetProvider getColumnOffsetProvider() {
        return columnOffsetProvider;
    }

    public PowerSwitcherType getPowerSwitcherType() {
        return powerSwitcherType;
    }

    public boolean isHasHotswap() {
        return hasHotswap;
    }

    public boolean isMagneticWristRestHolder() {
        return magneticWristRestHolder;
    }
	
	public int getLastCol(){
		return columnsCount - 1;
	}
	
	public int getLastRow(){
		return rowsCount - 1;
	}

    public double getBordersOffset() {
        return bordersOffset;
    }
}

File: cad3d/src/main/java/com/github/grishberg/cad3d/keyboard/cfg/KeyboardConfig.kt
package com.github.grishberg.cad3d.keyboard.cfg

import com.github.grishberg.cad3d.plugin.cfg.KeyboardPart
import com.github.grishberg.cad3d.plugin.Config
import com.github.grishberg.cad3d.plugin.cfg.KeyPlaceholderType
import com.github.grishberg.cad3d.plugin.cfg.PowerSwitcherType
import com.github.grishberg.cad3d.plugin.cfg.SettingsContainer
import com.github.grishberg.cad3d.plugin.cfg.ThumbClusterSettings
import com.github.grishberg.cad3d.plugin.cfg.TrackballConfig

data class KeyboardConfig(
    val fn: Int,
    val stlFn: Int,
    val plateZOffset: Double,
    val rowCurvature: Double,
    val tentingAngle: Double,
    val columnCurvature: Double,
    val keyswitchHeight: Double,
    val keyswitchWidth: Double,
    val extraWidth: Double,
    val extraHeight: Double,
    val plateThickness: Double,
    val saProfileKeyHeight: Double,
    val centerCol: Int,
    val rowsCount: Int,
    val columnsCount: Int,
    val centerRow: Int,
    val keyPlaceHolderWidth: Double,
    val keyPlaceHolderDepth: Double,
    val keyPlaceHolderHeight: Double,
    val isLowProfile: Boolean,
    val zAngleProvider: KeyZAngleProvider, //Не нужно сериализовывать
    val columnOffsetProvider: KeyOffsetProvider, //Не нужно сериализовывать
    val powerSwitcherType: PowerSwitcherType,
    val isHasHotswap: Boolean,
    val isMagneticWristRestHolder: Boolean,
    val bordersOffset: Double,
    val visibleKeyboardParts: Set<KeyboardPart>,
    val modifiedKeyboardParts: Set<KeyboardPart>,
    val thumbClusterSettings: ThumbClusterSettings,
    val screwNutHoleDiameter: Double,
    val screwHolderWallhickness: Double,
    val screwBoltDiameter: Double = 3.0,
    val isSkeletonMode: Boolean,
    val trackball: TrackballConfig,
    val wallsSettings: WallsSettings,
    val controllerPlateHeight: Double = 1.5,
    val keyPlaceholderType: KeyPlaceholderType,
    val horizontalExtraSpace: Double,
    val verticalExtraSpace: Double,
): Config {

    val lastCol: Int
        get() = columnsCount - 1
    val lastRow: Int
        get() = rowsCount - 1

    companion object {
        fun SettingsContainer.getKeyboardConfig(modifiedKeyboardParts: Set<KeyboardPart>): KeyboardConfig = KeyboardConfig(
            fn = keyboardSettings.fn,
            stlFn = keyboardSettings.stlFn,
            plateZOffset = keyboardSettings.plateZOffset,
            rowCurvature = keyboardSettings.rowCurvature,
            tentingAngle = keyboardSettings.tentingAngle,
            columnCurvature = keyboardSettings.columnCurvature,
            keyswitchHeight = 18.0,
            keyswitchWidth = 18.0,
            extraWidth = 2.5,
            extraHeight = 1.0,
            keyPlaceHolderWidth = 15.7,
            keyPlaceHolderDepth = 15.7,
            keyPlaceHolderHeight = 4.0,
            horizontalExtraSpace = keyboardSettings.horizontalExtraSpace,
            verticalExtraSpace = keyboardSettings.verticalExtraSpace,
            zAngleProvider = KeyZAngleProvider(),
            columnOffsetProvider = KeyOffsetProvider(),
            plateThickness = keyboardSettings.plateThickness,
            saProfileKeyHeight = keyboardSettings.saProfileKeyHeight,
            columnsCount = keyboardSettings.columnsCount,
            rowsCount = keyboardSettings.rowsCount,
            centerCol = keyboardSettings.centerCol,
            centerRow = keyboardSettings.centerRow,
            isLowProfile = keyboardSettings.isLowProfile,
            powerSwitcherType = keyboardSettings.powerSwitcherType,
            isHasHotswap = keyboardSettings.isHasHotswap,
            isMagneticWristRestHolder = keyboardSettings.isMagneticWristRestHolder,
            bordersOffset = keyboardSettings.bordersOffset,
            visibleKeyboardParts = assemblySettings.toKeyboardPartsList(),
            modifiedKeyboardParts = modifiedKeyboardParts,
            thumbClusterSettings = thumbClusterSettings,
            screwNutHoleDiameter = keyboardSettings.screwNutHoleDiameter,
            screwHolderWallhickness = keyboardSettings.screwHolderWallhickness,
            isSkeletonMode = keyboardSettings.isSkeletonMode,
            trackball = trackballSettings,
            wallsSettings = WallsSettings(),
            keyPlaceholderType = keyboardSettings.keyPlaceholderType,
        )
    }
}

